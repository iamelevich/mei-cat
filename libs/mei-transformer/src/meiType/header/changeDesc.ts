import { type Static, Type } from "@sinclair/typebox";
import { StandardTagSchema } from "../common";
import { AttrBiblSchema } from "../shared/attr/bibl";
import { AttrCommonSchema } from "../shared/attr/common";
import { AttrLangSchema } from "../shared/attr/lang";
import { PSchema } from "../shared/p";

/**
 * Description of a revision of the MEI file.
 * @see https://music-encoding.org/guidelines/v5/elements/changeDesc.html
 */
export const ChangeDescSchema = Type.Intersect([
	StandardTagSchema,
	AttrCommonSchema,
	AttrLangSchema,
	AttrBiblSchema,
	Type.Object({
		/**
		 * One or more text phrases that form a logical prose passage (optional).
		 * @see https://music-encoding.org/guidelines/v5/elements/p.html
		 */
		p: Type.Optional(Type.Union([PSchema, Type.Array(PSchema)])),
	}),
]);

export type ChangeDesc = Static<typeof ChangeDescSchema>;
