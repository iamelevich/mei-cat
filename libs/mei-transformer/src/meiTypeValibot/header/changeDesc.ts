import * as v from "valibot";
import { StandardTagSchema } from "../common";
import { AttrCommonSchema } from "../shared/attr/common";
import { AttrLangSchema } from "../shared/attr/lang";

/**
 * Simplified schema for changeDesc element.
 * @see https://music-encoding.org/guidelines/v5/elements/changeDesc.html
 */
export const ChangeDescSchema = v.object({
	...StandardTagSchema.entries,
	...AttrCommonSchema.entries,
	...AttrLangSchema.entries,
});

export type ChangeDescData = v.InferOutput<typeof ChangeDescSchema>;
