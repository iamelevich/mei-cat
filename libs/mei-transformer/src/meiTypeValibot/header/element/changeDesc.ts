import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import {
	AttrBiblSchema,
	AttrCommonSchema,
	AttrLangSchema,
	ModelPLikeSchema,
} from "../../shared";

/**
 * Base schema with attribute, to simplify types for ChangeDescSchema
 */
const ChangeDescBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrLangSchema.entries,
});

/**
 * Description of a revision of the MEI file.
 * @see https://music-encoding.org/guidelines/v5/elements/changeDesc.html
 */
export const ChangeDescSchema = v.intersect([
	ChangeDescBaseSchema,
	ModelPLikeSchema,
]);

export type ChangeDescData = v.InferOutput<typeof ChangeDescSchema>;
