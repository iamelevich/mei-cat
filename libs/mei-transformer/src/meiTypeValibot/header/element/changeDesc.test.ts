import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { ChangeDescSchema } from "./changeDesc";

describe("ChangeDescSchema", () => {
	it("should be defined", () => {
		expect(ChangeDescSchema).toBeDefined();
	});

	it("should be a valid schema", () => {
		const parseResult = v.safeParse(ChangeDescSchema, {
			p: {
				"#text": "tab 1",
				"@xml:id": "p_N25CCC",
			},
			"@xml:id": "changeDesc_N25CCB",
		});

		expect(parseResult.success).toBe(true);
		expect(parseResult.output).toMatchSnapshot();
	});
});
