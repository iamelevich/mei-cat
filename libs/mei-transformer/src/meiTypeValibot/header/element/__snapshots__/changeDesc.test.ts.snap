// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ChangeDescSchema should be a valid schema 1`] = `
{
  "@xml:id": "changeDesc_N25CCB",
  "p": {
    "#text": "tab 1",
    "@xml:id": "p_N25CCC",
  },
}
`;
